### PR TITLE
feat: as a Portal, I can create an Attestation

### LINKED ISSUE #61: [US] As a Portal SC, I can create an Attestation on the AttestationRegistry
As a Portal, I can anchor on chain an Attestation. Only registered Portals can anchor an Attestation on-chain
More details on this specification in the issue #7 

## Requirements
* Only registered Portals can create Attestations
* Attestations are created through the `Attestation Registry` smart contract
* Attestations are identified via their keccack(portalId, nonce), where the nonce is the number of attestations already created with this Portal, minus 1. (The third Attestation created with the Portal `0xabc...ef` will have the id keccack(abi.encodePacked(`0xabc...ef`, 1)
* Attestations should be found:
           * by index, for a specific Portal
           * directly by their id
* Creating an Attestation should trigger an event

## Todo
- [ ] Create a `AttestationRegistry` contract
- [ ] Create a `Attestation` struct to store a Attestation data
- [ ] Add a mapping to store Attestations per Portal, per index (portalAddress => AttestationIndex => struct Attestation)
- [ ] Add a mapping to store Attestations per Id (AttestationId => tuple(portalAddress, AttestationIndex))
- [ ] Add an mapping of uint256 to get the number of Attestations per portalAddress
- [ ] Add a getter to get the length of an array of Attestations for a Portal
- [ ] Add a `attest(bytes32 schemaId, bytes attestationData)` function to add a new Attestation to the mapping and the array. This create method is only callable by a Portal registered. Add also the call to this method in the DefaultPortal
- [ ] Add a `revoke(bytes32 attestationId)` function to add a new Attestation to the mapping and the array. This create method is only callable by a Portal registered. Add also the call to this method in the DefaultPortal
- [ ] To create an Attestation, it must have `bytes data` and `bytes32 schemaId`
- [ ] When an Attestation is created, 1/ add it to the attestationsPerPortal 2/ add its portal and index in the list , 2/ add it to the 3/ emit a custom `AttestationCreated` event
- [ ] Add a `AttestationRegistry` script to be able to deploy it easily + a pnpm command to trigger it
- [ ] Add a `AttestationRegistry` set of unit tests to cover the happy path + the errors and events triggering

## Naive implementation of the AttestationRegistry
```

contract AttestationRegistry()

{

	public mapping (address => uint256) attestationsLengthPerPortal;
	public mapping (address => Attestation[]) attestationsPerPortal;
	public mapping (bytes32 => uint256) attestations;
	

	modifier onlyPortals()
	{
		// @dev: Test that the portalRegistry knows this Portal
		if (bytes(portalRegistry.portals(msg.sender)) == '0x')
		{
			revert AttestationCreatorNotAPortal(msg.sender);
		}
		_;
	}
	

	function attest(bytes32 schemaId, bytes attestationData) onlyPortals()
	{

		bytes32 attestationId = keccack(abi.encodePacked(msg.sender, attestationsLengthPerPortal[msg.sender]));

		Attestation attestation = Attestation(
			schemaId: schemaId,
			attestationData: attestationData,
			isRevoked: false,
			createdAt: block.timestamp
			attestationId: attestationId
		)

		attestationsPerPortal.push(attestation);
		attestations[attestationId] = attestation;
		attestationsLengthPerPortal[msg.sender]++;
                emit AttestationCreated(schemaId, msg.sender, attestationId, attestationData)
	}


}


```

## Acceptance criteria
* Should include a script to deploy the `Attestation Registry`
* Should include 100% unit-tests (including accessing the metadata through the mapping)

## Additional information
| Workload estimate | 
|--------|
| TBD |

### PR DESCRIPTION
**Pull request by 0xEillo**
## What does this PR do?

As a Portal SC, I can create an Attestation on the AttestationRegistry. 

Some slight changes to the US. Noteably, in this implementation I assume that it is the Portal's role to construct the Attestation, thereby the attest function directly receives an "Attestation" as a parameter. 

### Related ticket

Fixes #61 and #33

### Type of change

- [ ] Chore
- [ ] Bug fix
- [x] New feature
- [ ] Documentation update

## Check list

- [x] Unit tests for any smart contract change
- [x] Contracts and functions are documented

### CODE REVIEW

File: src/types/Structs.sol
Comment (orbmis) [inline src/types/Structs.sol:1]: There seems to be a conflict of conventions here as we have this file and we also have a "structs" folder, and many of the Structs are declared in both places.   It doesn't really matter which convention we use we should pick one and remove the other.

**Comment by satyajeetkolhapure** [comment]
Currently we have a default portal which has a schema and list of modules. Attestation struct hold schema too. We need to decide which component owns the schema.

**Comment by satyajeetkolhapure** [comment]
> I think its fine to have a reference to the schemaId in both don't you think?

I can be useful but then there will be extra validations required like portal and attestation having different schema id situation. Also it can be little confusing while implementing portals. Do we have any strong reason to do so?

**Comment by alainncls** [comment]
@0xEillo let's push the tests under `test/unit` and leave the mocks where they are.
This way, by modifying the command in `smart-contracts-test.yml`, we'll be able to get the coverage we want (only on the real tests and not the mocks).

File: src/AttestationRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  /**
+   * @notice Revokes an attestation of given identifier
+   * @param attestationId the attestation identifier
+   */
+  function revoke(bytes32 attestationId) external {
+    if (!isRegistered(attestationId)) revert AttestationNotAttested();
+    if (msg.sender != attestations[attestationId].portal) revert OnlyAttestingPortal();
+
+    attestations[attestationId].revoked = true;
```
[the remark this answers is not in the document]
Comment (orbmis) [reply src/AttestationRegistry.sol:78]: Removing this comment because this is being addressed by another story

**Review by orbmis — CHANGES_REQUESTED** [review]

File: test/AttestationRegistry.t.sol [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    AttestationRegistry testAttestationRegistry = new AttestationRegistry();
+
+    vm.expectRevert(AttestationRegistry.PortalRegistryInvalid.selector);
+    testAttestationRegistry.initialize(address(0), schemaRegistryAddress);
+
+    vm.expectRevert(AttestationRegistry.SchemaRegistryInvalid.selector);
+    testAttestationRegistry.initialize(portalRegistryAddress, address(0));
+  }
+
+  function test_attest(Attestation memory attestation) public {
```
[the remark this answers is not in the document]
Comment (0xEillo) [reply test/AttestationRegistry.t.sol:52]: Not a dumb question. In foundry you can have parameters to test functions. Foundry will then create thousands of iterations of an attestation and test that they all work. i.e Fuzzing

File: src/AttestationRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    schemaRegistry = SchemaRegistry(_schemaRegistry);
+  }
+
+  /**
+   * @notice Registers an attestation to the AttestationRegistry
+   * @param attestation the attestation to register
+   * @dev This method is only callable by a registered Portal
+   */
+  function attest(Attestation calldata attestation) external onlyPortals(msg.sender) {
+    attestations[attestation.attestationId] = attestation;
```
[the remark this answers is not in the document]
Comment (0xEillo) [reply src/AttestationRegistry.sol:66]: Yes, we can do this inside the Abstract Portal where the attestation will be created

[the remark this answers is not in the document]
Comment (0xEillo) [reply src/AttestationRegistry.sol:66]: We can do this check in the Portal as well probably

**Review by alainncls — APPROVED** [review]

**Pull request merged**